sources/core.ts

```diff
@@ -491,37 +491,53 @@ export function cloneNode(input: Node, offset: number = 0) {
 }
 
-export function registerDynamic(machine: StateMachine, from: number, test: Callback<keyof typeof tests>, to: number, reducer?: Callback<keyof typeof reducers>) {
-    machine.nodes[from].dynamics.push([test, {to, reducer}]);
+export function registerDynamic<T extends keyof typeof tests, R extends keyof typeof reducers>(machine: StateMachine, from: number, test: Callback<T, typeof tests>, to: number, reducer?: Callback<R, typeof reducers>) {
+    machine.nodes[from].dynamics.push([
+        test as Callback<keyof typeof tests, typeof tests>,
+        {to, reducer: reducer as Callback<keyof typeof reducers, typeof reducers>}
+    ]);
 }
 
-export function registerShortcut(machine: StateMachine, from: number, to: number, reducer?: Callback<keyof typeof reducers>) {
-    machine.nodes[from].shortcuts.push({to, reducer});
+export function registerShortcut<R extends keyof typeof reducers>(machine: StateMachine, from: number, to: number, reducer?: Callback<R, typeof reducers>) {
+    machine.nodes[from].shortcuts.push(
+        {to, reducer: reducer as Callback<keyof typeof reducers, typeof reducers>}
+    );
 }
 
-export function registerStatic(machine: StateMachine, from: number, test: string, to: number, reducer?: Callback<keyof typeof reducers>) {
+export function registerStatic<R extends keyof typeof reducers>(machine: StateMachine, from: number, test: string, to: number, reducer?: Callback<R, typeof reducers>) {
     let store = !Object.prototype.hasOwnProperty.call(machine.nodes[from].statics, test)
         ? machine.nodes[from].statics[test] = []
         : machine.nodes[from].statics[test];
 
-    store.push({to, reducer});
+    store.push({to, reducer: reducer as Callback<keyof typeof reducers, typeof reducers>});
 }
 
 // ------------------------------------------------------------------------
 
-export type CallbackStore<T extends string> = {[key: string]: (state: RunState, segment: string, ...args: any[]) => {}};
-export type Callback<T extends string> = T | [T, ...any[]];
-
-export function execute<T extends string>(store: CallbackStore<T>, callback: Callback<T>, state: RunState, segment: string): any {
+type UndefinedKeys<T> = {[P in keyof T]-?: undefined extends T[P] ? P : never}[keyof T];
+type UndefinedTupleKeys<T extends unknown[]> = UndefinedKeys<Omit<T, keyof []>>;
+type TupleKeys<T> = Exclude<keyof T, keyof []>;
+
+export type CallbackFn<P extends any[], R> = (state: RunState, segment: string, ...args: P) => R;
+export type CallbackFnParameters<T extends CallbackFn<any, any>> = T extends ((state: RunState, segment: string, ...args: infer P) => any) ? P : never;
+export type CallbackStore<T extends string, R> = Record<T, CallbackFn<any, R>>;
+export type Callback<T extends string, S extends CallbackStore<T, any>> =
+    [TupleKeys<CallbackFnParameters<S[T]>>] extends [UndefinedTupleKeys<CallbackFnParameters<S[T]>>]
+        ? (T | [T, ...CallbackFnParameters<S[T]>])
+        : [T, ...CallbackFnParameters<S[T]>];
```

Comment:
That's ... hard to read 😅 I wonder if it should be in its own file, with more newlines and comments.

Reply:
Maybe for another time 😄